File: src/IRC/Channel.ts
/**
 * Channel — IVC channel object (#channel / &channel).
 *
 * Mirrors PHP `Fortress\IRC\Serv\ChanServ` object semantics.
 *
 * Usage:
 *   const chan = await Channel.fromServer('#fortress');
 *   const isOp = chan.hasMode('o');
 *   const access = await chan.checkAccess('CyberFox');
 */

import { IrcObject } from './IvcObject.js';
import { IvcClient }    from './IvcClient.js';
import { IvcMarshaller } from './IvcMarshaller.js';
import type { IvcAccessResult, IvcApiResponse, IvcResponse } from './types.js';

export class Channel extends IrcObject {
  constructor(private readonly _name: string) {
    super();
  }

  get target(): string {
    return this._name;
  }

  get name(): string {
    return this._name;
  }

  // ---------------------------------------------------------------
  // Channel-specific helpers
  // ---------------------------------------------------------------

  /**
   * Normalise a channel name — ensures a leading '#'.
   */
  static normalize(name: string): string {
    name = name.trim();
    if (!name.startsWith('#') && !name.startsWith('&')) {
      name = '#' + name;
    }
    return name;
  }

  /**
   * Check access for a user against this channel's mode flags.
   * Delegates to the server's /api/access.php endpoint.
   */
  async checkAccess(user?: string): Promise<IvcAccessResult> {
    const result = await IvcClient.checkAccess(this._name, user);
    return result.body as IvcAccessResult;
  }

  /**
   * True if the locally cached modes include +o (operator flag).
   * For user-specific op status, use checkAccess().
   */
  get isSecret(): boolean   { return this.hasMode('s'); }
  get isModerated(): boolean { return this.hasMode('m'); }
  get isTopicLocked(): boolean { return this.hasMode('t'); }
  get isInviteOnly(): boolean { return this.hasMode('i'); }
  get passkey(): string | null { return this.getProp('k'); }

  // ---------------------------------------------------------------
  // Factory / auto-marshalling
  // ---------------------------------------------------------------

  static fromBody(body: IvcApiResponse): Channel {
    const name = (body.base_target as string | undefined) ?? '';
    const chan  = new Channel(name);
    return IrcObject._applyBody(chan, body);
  }

  /**
   * Fetch a channel object from the server and return a hydrated Channel.
   *
   *   const chan = await Channel.fromServer('#fortress');
   *   console.log(chan.isModerated);
   */
  static async fromServer(name: string): Promise<Channel> {
    name = Channel.normalize(name);
    const result = await IvcClient.fetchObject(name);
    const chan   = Channel.fromBody({ ...result.body, base_target: name });
    if (result.status) chan._lastStatus = result.status;
    return chan;
  }
}

// Register '# ' and '&' prefixes with the marshaller.
IvcMarshaller.register('#', Channel);
IvcMarshaller.register('&', Channel);

File: ivc-client/src/main/java/ivc/Channel.java
package ivc;

import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * Channel — IVC channel object (#channel / &channel).
 *
 * Mirrors PHP Fortress\IRC\Serv\ChanServ object semantics.
 */
public class Channel extends IrcObject {

    private final String name;

    public Channel(String name) {
        this.name = normalize(name);
    }

    @Override
    public String target() { return name; }

    public String getName() { return name; }

    public static String normalize(String name) {
        name = name == null ? "" : name.trim();
        if (!name.startsWith("#") && !name.startsWith("&")) name = "#" + name;
        return name;
    }

    // --- Mode flag helpers ---

    public boolean isSecret()      { return hasMode("s"); }
    public boolean isModerated()   { return hasMode("m"); }
    public boolean isTopicLocked() { return hasMode("t"); }
    public boolean isInviteOnly()  { return hasMode("i"); }
    public String  passkey()       { return getProp("k"); }

    /** Check access for a user via the server. */
    public CompletableFuture<Map<String, Object>> checkAccess(String user) {
        return IvcClient.checkAccess(name, user).thenApply(r -> r.body());
    }

    // --- Factory / auto-marshalling ---

    public static Channel fromBody(Map<String, Object> body) {
        String bt  = body.getOrDefault("base_target", "").toString();
        Channel ch = new Channel(bt.isEmpty() ? "#unknown" : bt);
        return IrcObject.applyBody(ch, body);
    }

    /**
     * Fetch a channel object from the server and return a hydrated Channel.
     *
     *   Channel chan = Channel.fromServer("#fortress").join();
     *   System.out.println(chan.isModerated());
     */
    public static CompletableFuture<Channel> fromServer(String name) {
        String norm = normalize(name);
        return IvcClient.fetchObject(norm).thenApply(r -> {
            Map<String, Object> body = new java.util.LinkedHashMap<>(r.body());
            body.putIfAbsent("base_target", norm);
            Channel ch = fromBody(body);
            if (r.status() != null) ch.lastStatus = r.status();
            return ch;
        });
    }
}